include/cache_guttering.h

```diff
@@ -21,32 +21,29 @@ class CacheGuttering : public GutteringSystem {
   static constexpr size_t l1_cache_size   = 32768;    // l1 cache bytes per cpu
   static constexpr size_t l2_cache_size   = 1048576;  // l2 cache bytes per cpu
   static constexpr size_t l3_cache_size   = 33554432; // l3 cache bytes in total
+
+  static constexpr size_t cache_sizes[3] = {32768, 1048576, 33554432};
   static constexpr size_t cache_line      = 64; // number of bytes in a cache_line
-  static constexpr size_t cache_bytes_per_child = 2 * cache_line;
-  static constexpr size_t RAM_bytes_per_child   = 8 * cache_line;
-
-  // basic 'tree' params, hardcoded for now. TODO: Determine by sizes later
-  static constexpr size_t fanout          = 32;  // fanout for cache gutters. Must be power of 2
-  static constexpr size_t num_l1_bufs     = 8;   // number l1 buffers. Must be power of 2
-  static constexpr size_t num_l2_bufs     = num_l1_bufs * fanout;
-  static constexpr size_t num_l3_bufs     = num_l2_bufs * fanout;
-  static constexpr size_t max_RAM1_bufs   = num_l3_bufs * fanout;
-  static constexpr size_t l1l2buffer_elms = cache_bytes_per_child * fanout / sizeof(update_t);
-  static constexpr size_t l3buffer_elms   = cache_bytes_per_child * fanout / sizeof(update_t);
-
-  // bit length variables
-  static constexpr int l1_bits   = log2_constexpr(num_l1_bufs);
-  static constexpr int l2_bits   = log2_constexpr(num_l2_bufs);
-  static constexpr int l3_bits   = log2_constexpr(num_l3_bufs);
-  static constexpr int RAM1_bits = log2_constexpr(max_RAM1_bufs);
+  static constexpr size_t numerator         = 1;
+  static constexpr size_t divisor         = 1;
+  
+  static constexpr size_t buf_bytes[4] = {2*cache_line, 4*cache_line, 8*cache_line, 16*cache_line};
+  static constexpr size_t buf_elems[4] = {buf_bytes[0] / sizeof(update_t), buf_bytes[1] / sizeof(update_t),
+    buf_bytes[2] / sizeof(update_t), buf_bytes[3] / sizeof(update_t)};
+
+  static constexpr size_t num_bufs[3] = {numerator * cache_sizes[0] / (buf_bytes[0] * divisor), numerator * cache_sizes[1] / (buf_bytes[1] * divisor), 
+    numerator * cache_sizes[2] / (buf_bytes[2] * divisor)};
+  static constexpr size_t max_RAM1_bufs = num_bufs[2] * 32;
+  
+  static constexpr int bits[4] = {log2_constexpr(num_bufs[0]), log2_constexpr(num_bufs[1]), 
+    log2_constexpr(num_bufs[2]), log2_constexpr(max_RAM1_bufs)};
```

Comment:
So I don't like these changes for a couple reasons.
First, I think they reduce the readability of the code. In my opinion it would be better to have very descriptive variables that we can also put into arrays so we can put the flushes in one function. Similar to what you did for the cache sizes. For example, num_bufs[3] = {num_l1_bufs, num_l2_bufs, num_l3_bufs}.

Second, these changes mean that the amount of data written from each node to one of its children is now on average 8 bytes. For example, the number of L1 buffers is 32768 / 128 = 256 with 128 bytes per buffer, therefore, with a fanout of 16 each child gets on average 8 bytes.

How do these changes affect performance?

Reply:
Just ran a quick test. For Kron17 our performance is only 220million/s with 20 threads which is a reduction of 100 million/s. It should be pretty easy for us to fix this by adjusting some of the parameters.

Reply:
I think both approaches have readability and usability concerns; we can iron out readability over time. The previous iteration over allocated memory, because we focused on bytes_per_child. This iteration allocates the right amount of memory, but it's much easier to have very few bytes_per_child. We'll have to have some sort of sanity check or better math.

Reply:
Agreed, let's talk it soon. Sanity checks with better math are always good.